src/ucp/wireup/select.c

```diff
@@ -1618,55 +1648,54 @@ ucp_wireup_add_bw_lanes(const ucp_wireup_select_params_t *select_params,
             dev_index         = context->tl_rscs[rsc_index].dev_index;
-            sinfo->path_index = dev_count.local[dev_index];
+            sinfo->path_index = dev_count->local[dev_index];
 
         } else {
             /* disqualify/count lane_desc_idx */
-            addr_index      = select_ctx->lane_descs[excl_lane].addr_index;
-            rsc_index       = select_ctx->lane_descs[excl_lane].rsc_index;
-            dev_index       = context->tl_rscs[rsc_index].dev_index;
-            excl_lane       = UCP_NULL_LANE;
-            skip_dev_index  = dev_index;
+            addr_index        = select_ctx->lane_descs[dev_count->excl_lane].addr_index;
+            rsc_index         = select_ctx->lane_descs[dev_count->excl_lane].rsc_index;
+            dev_index         = context->tl_rscs[rsc_index].dev_index;
+            dev_count->excl_lane      = UCP_NULL_LANE;
+            dev_count->skip_dev_index = dev_index;
         }
 
         /* Count how many times the LOCAL device is used */
         iface_attr = ucp_worker_iface_get_attr(ep->worker, rsc_index);
-        ++dev_count.local[dev_index];
+        ++dev_count->local[dev_index];
 
         /* Count how many times the REMOTE device is used */
         ae = &select_params->address->address_list[addr_index];
-        ++dev_count.remote[ae->dev_index];
+        ++dev_count->remote[ae->dev_index];
 
         /* Account for possible path override */
         local_num_paths  = iface_attr->dev_num_paths;
         remote_num_paths = ae->dev_num_paths;
         if (allow_extra_path &&
-            ((skip_dev_index != UCP_NULL_RESOURCE) && /* clang sanitizer */
-             (skip_dev_index == dev_index))) {
+            ((dev_count->skip_dev_index != UCP_NULL_RESOURCE) && /* clang sanitizer */
+             (dev_count->skip_dev_index == dev_index))) {
             /* Allow path since we skipped one */
             local_num_paths++;
             remote_num_paths++;
 
             /* Remove that skipped path from score computation */
-            dev_count.local_skip[dev_index]      = 1;
-            dev_count.remote_skip[ae->dev_index] = 1;
+            dev_count->local_skip[dev_index]      = 1;
+            dev_count->remote_skip[ae->dev_index] = 1;
         }
 
-        if (dev_count.local[dev_index] >= local_num_paths) {
+        if (dev_count->local[dev_index] >= local_num_paths) {
             /* exclude local device if reached max concurrency level */
-            local_dev_bitmap &= ~UCS_BIT(dev_index);
+            bw_info->local_dev_bitmap &= ~UCS_BIT(dev_index);
         }
 
         /* Assume symmetric num_paths */
-        if (dev_count.remote[ae->dev_index] >= remote_num_paths) {
+        if (dev_count->remote[ae->dev_index] >= remote_num_paths) {
             /* exclude remote device if reached max concurrency level */
-            remote_dev_bitmap &= ~UCS_BIT(ae->dev_index);
+            bw_info->remote_dev_bitmap &= ~UCS_BIT(ae->dev_index);
```

Comment:
i think `local_dev_bitmap` and `remote_dev_bit_map` also has to be tracked by mem_type. with this change, a given mem_type can now consume paths of another mem_type, since `bw_info` is global to all mem_types.

Reply:
yeah maybe we can just add array of struct called `ucp_wireup_add_bw_lanes_state_t` instantiated per mem_type above `ucp_wireup_add_bw_lane()` and move dev_count and others inside.

Reply:
Yes, please see my latest commit and let me know.